src/reducers/__tests__/existingThoughtChange.js

```diff
@@ -10,27 +10,176 @@ it('edit a thought', () => {
+  expect(getContexts(stateNew, 'aa1'))
+    .toMatchObject([{
+      context: ['a'],
+      rank: 0,
+    }])
+  expect(getThoughts(stateNew, ['a']))
+    .toMatchObject([{ value: 'aa1', rank: 0 }])
+
+})
+
+it('edit a thought with descendants', () => {
+
+  const steps = [
+    newThought({ value: 'a' }),
+    newThought({ value: 'a1', insertNewSubthought: true }),
+    newThought({ value: 'a2' }),
+    existingThoughtChange({
+      newValue: 'aa',
+      oldValue: 'a',
+      context: [ROOT_TOKEN],
+      thoughtsRanked: [{ value: 'a', rank: 0 }]
+    })
+  ]
+
+  // run steps through reducer flow and export as plaintext for readable test
+  const stateNew = reducerFlow(steps)(initialState())
+  const exported = exportContext(stateNew, [ROOT_TOKEN], 'text/plaintext')
+
+  expect(exported).toBe(`- ${ROOT_TOKEN}
+  - aa
+    - a1
+    - a2`)
+
+  // aa should exist in ROOT context
+  expect(getContexts(stateNew, 'aa'))
+    .toMatchObject([{
+      context: [ROOT_TOKEN]
+    }])
+  expect(getThoughts(stateNew, ['aa']))
+    .toMatchObject([{ value: 'a1', rank: 0 }, { value: 'a2', rank: 1 }])
+
+})
+
+it('edit a thought existing in mutliple contexts', () => {
+
+  const steps = [
+    newThought({ value: 'a' }),
+    newThought({ value: 'ab', insertNewSubthought: true }),
+    newThought({ value: 'b', at: [{ value: 'a', rank: 0 }] }),
+    newThought({ value: 'ab', insertNewSubthought: true }),
+    existingThoughtChange({
+      newValue: 'abc',
+      oldValue: 'ab',
+      context: ['a'],
+      thoughtsRanked: [{ value: 'a', rank: 0 }]
+    })
+  ]
+
+  // run steps through reducer flow and export as plaintext for readable test
+  const stateNew = reducerFlow(steps)(initialState())
+  const exported = exportContext(stateNew, [ROOT_TOKEN], 'text/plaintext')
+
+  expect(exported).toBe(`- ${ROOT_TOKEN}
+  - a
+    - abc
+  - b
+    - ab`)
+
+  // abc should exist in context a
+  expect(getContexts(stateNew, 'abc'))
+    .toMatchObject([{
+      context: ['a']
+    }])
+  expect(getThoughts(stateNew, ['a']))
+    .toMatchObject([{ value: 'abc', rank: 0 }])
+
+})
+
+it('edit a thought that exists in another context', () => {
+
+  const steps = [
+    newThought({ value: 'a' }),
+    newThought({ value: 'ab', insertNewSubthought: true }),
+    newThought({ value: 'b', at: [{ value: 'a', rank: 0 }] }),
+    newThought({ value: 'a', insertNewSubthought: true }),
+    existingThoughtChange({
+      newValue: 'ab',
+      oldValue: 'a',
+      context: ['b'],
+      thoughtsRanked: [{ value: 'b', rank: 1 }, { value: 'a', rank: 0 }]
+    })
+  ]
+
+  // run steps through reducer flow and export as plaintext for readable test
+  const stateNew = reducerFlow(steps)(initialState())
+  const exported = exportContext(stateNew, [ROOT_TOKEN], 'text/plaintext')
+
+  expect(exported).toBe(`- ${ROOT_TOKEN}
+  - a
+    - ab
+  - b
+    - ab`)
+
+  // ab should exist in both contexts a and b
+  expect(getContexts(stateNew, 'ab'))
```

Comment:
`getContexts` return the context b with `rank: 0`.

Reply:
@raineorshine 
`getContexts` in this case returns the following:
`[
      {
        context: [ 'a' ],
        id: '266e7ff0-aa5c-4927-a8a2-bcc2591b8e85',
        rank: 0
      },
      {
        context: [ 'b' ],
        rank: 0,
        id: '0783bcd8-c2b8-4e48-9487-a530b3b44be2'
      }
    ]`
Notice that the `b` should have `rank: 1`, but instead it returns `rank: 0`, which is not desirable.

Reply:
This is correct actually. Since `getContexts` returns `ThoughtContexts` not `Child` nodes, `rank` represents the rank of the value within each context. i.e. in `Context` `['a']` the thought `ab` has rank `0`, and in `Context` `['b']` the thought `ab` also has rank `0`. Ranks are only unique within the same `Context`.